code/go/internal/validator/semantic/validate_kibana_filter_present.go

```diff
@@ -52,13 +53,74 @@ func ValidateKibanaFilterPresent(fsys fspath.FS) specerrors.ValidationErrors {
 }
 
 func checkDashboardHasFilter(file pkgpath.File) error {
+	err := findPanelsFilters(file)
+	if err != nil {
+		dashboardErr := findDashboardFilter(file)
+		if dashboardErr != nil {
+			return fmt.Errorf("%w and %w", dashboardErr, err)
+		}
+	}
+	return nil
+}
+
+func findPanelsFilters(file pkgpath.File) error {
+	panelsJSON, err := file.Values("$.attributes.panelsJSON")
+	if err != nil {
+		return fmt.Errorf("unable to find panels definition: %w", err)
+	}
+
+	var panels []struct {
+		EmbeddableConfig struct {
+			Attributes struct {
+				State struct {
+					Filters []any `mapstructure:"filters"`
+					Query   struct {
+						Query string `mapstructure:"query"`
+					} `mapstructure:"query"`
+				} `mapstructure:"state"`
+			} `mapstructure:"attributes"`
+		} `mapstructure:"embeddableConfig"`
+		Type string `mapstructure:"type"`
+	}
+	switch panelsJSON := panelsJSON.(type) {
+	case []any:
+		// Dashboard is decoded, as in source packages.
+		err = mapstructure.Decode(panelsJSON, &panels)
+		if err != nil {
+			return fmt.Errorf("unable to decode panels definition: %w", err)
+		}
+	case string:
+		// Dashboard is encoded as in built packages.
+		err = json.Unmarshal([]byte(panelsJSON), &panels)
+		if err != nil {
+			return fmt.Errorf("unable to decode embedded panels definition: %w", err)
+		}
+	default:
+		return fmt.Errorf("unexpected type for panels JSON: %T", panelsJSON)
+	}
+
+	for _, panel := range panels {
+		if panel.Type == "search" {
```

Comment:
@drewdaemon are there other types of panels for which it would be ok if they don't have a filter?

(this is the panel why I added this https://github.com/elastic/package-spec/pull/687/files#diff-0e0b0481bb49a2f2675e5a87302801e3489b686114aef00040cae507f211a011R604)

Reply:
Unfortunately, it's a bit more nuanced than panel type. For example, most panels of type `visualization` make ES requests and would benefit from adding filters. But, markdown panels, which don't make any requests and don't support filters, are also of type `visualization`.

We had to deal with similar nuances/edge cases to reliably [detect legacy visualizations](https://github.com/elastic/kbncontent/blob/main/kbncontent.go#L77-L97).

I'm thinking that `kbncontent` might be the place to put this logic since there's already a groundwork there. Maybe we could add a `CanUseFilter()` method to the [visualization descriptor struct](https://github.com/elastic/kbncontent/blob/main/kbncontent.go#L19-L25) and use it here?

Reply:
Good point, opened PR in https://github.com/elastic/kbncontent/pull/4, lets discuss there what visualizations we should check.